### ISSUE TITLE
Update to support official Nest Cloud API

### ISSUE THREAD

**Issue by KraigM** [KraigM/homebridge-nest#11]
Now that Nest has an **official** api, we need to switch to it so that we are properly following their usage.  Should solve many compatibility issues and we won't have to worry about losing support in the future.

**Comment by KraigM** [comment]
Has been merged into develop branch.  Hasn't been released yet (want to have willing volunteers try it out before I release it to everyone), but to any who want to try out the prerelease:

1. Install prerelease version 
`npm install -g git+https://github.com/KraigM/homebridge-nest.git#083e416c7db4d9a731ed15289ed95a2affe13aa8`

2. If you want to try the new API (not just fall back to the legacy code on the new version), follow guide to setup the new API by following the updated [README](https://github.com/KraigM/homebridge-nest/blob/develop/README.md)

**Comment by KraigM** [comment]
@brushj Did you update to git+https://github.com/KraigM/homebridge-nest.git#083e416c7db4d9a731ed15289ed95a2affe13aa8 ?

**Comment by KraigM** [comment]
@brushj Once you have pulled in the new version and followed the readme, you should find a warning line (likely yellow) in the log. It should look something like:

`[NestPlat] CODE IS ONLY VALID ONCE! Update config to use {'token':'c.ezX4zdQlPfEJsCtfn...WnvLK'} instead.`

of which you will take the `c.ezX4zdQlPfEJsCtfn...WnvLK` part to use as the token in the config

**Comment by brushj** [comment]
I must have been working on it right as you were updating everything. I
just ran the update but I am not getting anything when running homebridge
for the first time. Do I need to have something in my config as a place
holder for now in the token slot? It is the only field I don't have in the
config.

I have everything else setup in the nest end and in the dev portal it is
showing that one device is using the API.

On Wed, Dec 16, 2015 at 8:32 PM, KraigM <notifications@github.com> wrote:

> @brushj <https://github.com/brushj> Once you have pulled in the new
> version and followed the readme, you should find a warning line (likely
> yellow) in the log. It should look something like:
>
> [NestPlat] CODE IS ONLY VALID ONCE! Update config to use
> {'token':'c.ezX4zdQlPfEJsCtfn...WnvLK'} instead.
>
> of which you will take the c.ezX4zdQlPfEJsCtfn...WnvLK part to use as the
> token in the config
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KraigM/homebridge-nest/issues/11#issuecomment-165309330>
> .
>



-- 

Justin Brush
Systems Administrator
Flagler County Public Schools

**Comment by KraigM** [comment]
@brushj Do you see any warnings at all on the startup log?

**Comment by KraigM** [comment]
Don't put the token param in the config until you have a valid token to put in it.

**Comment by brushj** [comment]
@KraigM I don't have anything with the token in there now and I have no errors when I startup homebridge. I am running it on an rPi 2 with raspbian, for clarification.

**Comment by KraigM** [comment]
Run `npm list -g` and you should (hopefully) find a line that says 

`homebridge-nest@0.1.1 (git+https://github.com/KraigM/homebridge-nest.git#083e416c7db4d9a731ed15289ed95a2affe13aa8)`

**Comment by KraigM** [comment]
You can also jump on #slack if you want more instant response :-)

**Comment by brushj** [comment]
Just hopped on slack.

**Comment by leoneleone** [comment]
Hey Kraig. I've updated to the New API version. As you know I'm actually interested in getting my Nest Cams to work with the nest plugin. Because I don't have a Nest Thermostat (they are incompatible with my home hvac system unfortunately), I don't see an AWAY toggle in HomeKit after setting up the new plugin.

I think the best option is to create a MotionSensor service type for the Nest Cam, as the motion detection capabilities would be the most useful for a HomeKit setup in the meantime, until Apple HomeKit officially supports security cameras. If the Cams can be detected as MotionSensors perhaps the Away toggle will eventually be visible in HomeKit.

I will try to create a NestMotionSensor Service Type and an associated nest-motionsensor-accessory.js file in the lib folder. I think I can use some of the 'Motion Sensor" code from the homebridge-isy-js plugin as a template. Hopefully this will work to get the Nest Cams recognised.

The earlier version (unofficial nest api) had support for only one Nest Thermostat I believe. Does your new API script support multiple Nest accessories of the same type now?

**Comment by mguaylambert** [comment]
Hey! Awesome work on this plugin!! Can't wait for the new official API! Little question aside, do you know how work the commands with Siri, and if everything work in French?

Also, do you mind redirecting me on Slack and follow the thread?

**Comment by KraigM** [comment]
@mguaylambert Not sure what you mean by "do you mind redirecting me on Slack and follow the thread".  It is an open slack you just have to sign up.  Once on there, you can message me directly or on the nest channel.

**Comment by KraigM** [comment]
I updated develop with some new changes @bheung and I have been working on.  To update use 
```
npm install -g git+https://github.com/KraigM/homebridge-nest.git#92b2fca950ab3513ce91a67aea49a51dcb8d28ba
``` 
instead

**Comment by respectTheCode** [comment]
I am getting a `[TypeError: Cannot read property 'thermostats' of undefined]` error with `1a16c6d9875e4c2124bf00ab5206638e7f43ddbd`

**Comment by amro** [comment]
Thanks for writing this. Just a heads up, with the latest commit isn't working for me (with Siri) w/ the new API. If I switch to the old API, it works. Not sure what the deal is there.

**Comment by ryck** [comment]
Just wanted to say that the official API branch has been working perfectly for me for the past week, great job! My only issue is that Siri keeps calling it Living Room (Nest) but that's a minor thing...

**Comment by amro** [comment]
@ryck You can set the name in the Nest app

**Comment by ryck** [comment]
@amro: well, in the nest app you can set the location and a label, and Siri uses both, unless I am doing something wrong...

**Comment by amro** [comment]
@ryck So I think you just helped me figure out my issue. With the new API, I have to say something like "set the upstairs upstairs temperature to 75 degrees" and it works. I'll try removing the additional description in the nest app. It didn't work this way with the old API. Probably some Siri+HomeKit quirk.

**Comment by amro** [comment]
@ryck No, you're exactly right. I'm going to try removing the label.

**Comment by KraigM** [comment]
You can also rename what Siri calls it using Insteon+ app. Mine was "Hallway (Nest)" so I renamed it "Nest". Now even though the plugin still calls it "Hallway (Nest)" i tell Siri "set the Nest to 75 degrees"

**Comment by amro** [comment]
Thanks. Trying that now. FYI, I'm seeing this in the log:

    [Nest] Setting low target temperature for Downstairs (Nest) to: 76
    [Nest] Target temperature for Downstairs (Nest) is: 76 F
    FIREBASE WARNING: set at /devices/thermostats/9Vbf_m2aue-Q_hTHIupPVkhq4G0kGHzc/target_temperature_low_f failed: %s is larger than %s

**Comment by ryck** [comment]
@KraigM: you are absolutely right! Now Siri says just Nest. The key here was the 'Siri Control' field in Insteon+.

**Comment by KraigM** [comment]
You can also call it"thermostat". For instance, if you say "set the thermostat to 75 degrees" Siri will change all thermostats to 75. 

If you want to specify certain ones, just include the name of the zone or room. For instance, if you have a thermostat in the zone "upstairs" and one in the zone "downstairs", then you can say "set the upstairs thermostat to 75 degrees" and Siri will change the upstairs and leave the downstairs alone.

**Comment by ryck** [comment]
I only have one Nest, so I always use 'Set the temperature to 20 degrees' or 'what's the temperature in the $name_of_the_room'

**Comment by amro** [comment]
Thanks. The universal case works consistently. Perhaps this is some issue of taking a while to sync the names to iCloud or something. I'll try again in a bit.

**Comment by KraigM** [comment]
Released on version 1.0.0
